### ISSUE TITLE
Race condition when releasing a package.

### ISSUE THREAD

**Issue by jenstroeger** [jenstroeger/python-package-template#355]
When our `release.yaml` creates a Release ot Github

https://github.com/jenstroeger/python-package-template/blob/f41b0e6a46061081370253b9ac17a1b62b5c85dc/.github/workflows/release.yaml#L122-L126

we enter a race condition where the [Github Release event](https://docs.github.com/en/developers/webhooks-and-events/events/github-event-types#releaseevent) fires _before_ this current workflow finishes. Which can become an issue because other actions that trigger on that event (like our [`release-notification.yaml`](https://github.com/jenstroeger/python-package-template/blob/main/.github/workflows/release-notifications.yaml)) may want to access the artifacts attached to this workflow… and that fails because this workflow is still running.

**Comment by behnazh** [comment]
> @behnazh what do you think?

Yes that makes sense. The `release-notification.yaml` should not trigger on `release` event.

**Comment by jenstroeger** [comment]
@acrobat888 and I also talked about splitting the [`release.yaml`](https://github.com/jenstroeger/python-package-template/blob/main/.github/workflows/release.yaml) into two separate actions:

1. one to check & test & bump on `push` to the `main` branch; and
2. one to build the release artifacts & publish a release on `workflow_run` of the bump.

(The `release.yaml` already kinda sorta contains two workflows anyway.)

Other post-publish actions can then trigger on the `workflow_run` event of 2. and do their thing, _and_ have access to the `artifact-ubuntu-latest-python-3.10` artifact attached to the publish workflow run.

**Comment by behnazh** [comment]
Sounds good. I can split it to two workflows and possibly simplify uploading the provenance.

**Comment by behnazh** [comment]
Thinking some more, it might be better to add the notification to `release.yaml`. It would make it large, but it also prevents non-deterministic issues.

**Comment by jenstroeger** [comment]
> […] it might be better to add the notification to `release.yaml`.

@behnazh that would mean to either

- inline the code of the [`release-notification.yaml`](https://github.com/jenstroeger/python-package-template/blob/main/.github/workflows/release-notifications.yaml) and the [`release-wiki.yaml`](https://github.com/jenstroeger/python-package-template/pull/224) workflows at the end of the [`release.yaml`](https://github.com/jenstroeger/python-package-template/blob/main/.github/workflows/release.yaml) workflow; or
- to make those two [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) and invoke them at the end of the `release.yaml` workflow?

> It would make it large, but it also prevents non-deterministic issues.

Yup, and _all_ other post-publish workflows would also be added at the end of `release.yaml` then… hmm 🤔 There’s no better way?

**Comment by brad-getpassport** [comment]
@behnazh I refactored a lot of the code to effectively have two paths through the system based on the branch.

Since workflow_run can only execute on the `main` branch, we need to effectively have different paths through the system.

So I have a `check.yaml` that covers develop & main and verifies the build works.  If it does and if it's running on the `develop` branch it then runs specific jobs for develop.

After check is done, then a `release` job runs on main that does additional work, and only after that job completes do we then send out notifications as we wait for the _**workflow_run**_ event vs. a `release event`.

**Comment by behnazh** [comment]
> to make those two [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) (just like our [build.yaml](https://github.com/jenstroeger/python-package-template/blob/main/.github/workflows/build.yaml)) and invoke them at the end of the release.yaml workflow?

@jenstroeger @brad-getpassport  Reusable workflows would be a good way to go and would require minimal changes.

**Comment by brad-getpassport** [comment]
@behnazh that's what I did and it mostly works with the stuff I am working on, I told @jenstroeger he can share with you our stuff when he gets back home
